src/spec/cli.cr

```diff
@@ -0,0 +1,148 @@
+require "./formatter"
+require "option_parser"
+
+module Spec
+  # :nodoc:
+  class CLI
+    # :nodoc:
+    class Options
+      property formatters = Array(Spec::Formatter).new
+      property locations = Array({String, Int32}).new
+
+      property? default_formatter : Spec::Formatter? = nil
+      property? fail_fast : Bool = false
+      property? line : Int32? = nil
+      property? no_color : Bool = false
+      property? pattern : String? = nil
+      property? slowest : Int32? = nil
+    end
+
+    getter options : Options
+
+    private getter argv : Array(String)
+    private getter parser = OptionParser.new
+    private getter stderr : IO
+    private getter stdout : IO
+
+    private getter? prepared : Bool = false
+
+    def initialize(@argv, *, @options = Options.new, @stderr = STDERR, @stdout = STDOUT)
+      setup
+    end
+
+    def prepare
+      return if prepared?
+
+      parser.parse(argv)
+
+      unless argv.empty?
+        terminate "Error: unknown argument '#{argv.first}'"
+      end
+
+      if ENV["SPEC_VERBOSE"]? == "1"
+        options.default_formatter = Spec::VerboseFormatter.new
+      end
+    ensure
+      @prepared = true
```

Comment:
But you haven't prepared if there's an exception.

Reply:
Good point, will correct.